File: Matrix/SpiralMatrix.java
// LeetCode 54 - Spiral Matrix
// Time Complexity: O(m * n) | Space Complexity: O(m * n)
import java.util.ArrayList;
import java.util.List;

public class SpiralMatrix {
    public List<Integer> spiralOrder(int[][] matrix) {
        int rows = matrix.length;
        int cols = matrix[0].length;
        int[] directions = {0, 1, 0, -1, 0};
        boolean[][] visited = new boolean[rows][cols];
        List<Integer> answer = new ArrayList<>();

        int row = 0;
        int col = 0;
        int direction = 0;

        for (int remaining = rows * cols; remaining > 0; remaining--) {
            answer.add(matrix[row][col]);
            visited[row][col] = true;

            int nextRow = row + directions[direction];
            int nextCol = col + directions[direction + 1];
            if (nextRow < 0 || nextRow >= rows || nextCol < 0 || nextCol >= cols || visited[nextRow][nextCol]) {
                direction = (direction + 1) % 4;
            }

            row += directions[direction];
            col += directions[direction + 1];
        }

        return answer;
    }
}

File: Matrix/SpiralMatrix.go
// LeetCode 54 - Spiral Matrix
// Time Complexity: O(m * n) | Space Complexity: O(m * n)
package main

func spiralOrder(matrix [][]int) []int {
	rows := len(matrix)
	cols := len(matrix[0])
	directions := []int{0, 1, 0, -1, 0}
	visited := make([][]bool, rows)
	answer := make([]int, 0, rows*cols)

	for row := 0; row < rows; row++ {
		visited[row] = make([]bool, cols)
	}

	row := 0
	col := 0
	direction := 0

	for remaining := rows * cols; remaining > 0; remaining-- {
		answer = append(answer, matrix[row][col])
		visited[row][col] = true

		nextRow := row + directions[direction]
		nextCol := col + directions[direction+1]
		if nextRow < 0 || nextRow >= rows || nextCol < 0 || nextCol >= cols || visited[nextRow][nextCol] {
			direction = (direction + 1) % 4
		}

		row += directions[direction]
		col += directions[direction+1]
	}

	return answer
}
